include/picongpu/plugins/transitionRadiation/Calculator.hpp

```diff
@@ -0,0 +1,238 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "Particle.hpp"
+
+
+namespace picongpu
+{
+namespace transitionRadiation
+{
+    using complex_X = pmacc::math::Complex< float_X >;
+    using complex_64 = pmacc::math::Complex< float_64 >;
+
+    /* Arbitrary margin which is necessary to prevent division by 0 error
+        * created by particles moving in the plane of the foil.
+        */
+    float_X const DIV_BY_ZERO_MINIMUM = 1.e-7;
+
+    /** Calculator class for calculation of transition radiation.
+     * 
+     * @param particleSet transitionRadiation::Particle to compute transition radiation for
+     * @param lookDirection vector with observation direction
+     */
+    class Calculator
+    {
+
+    private:
+        transitionRadiation::Particle const & particle;
+        float3_X const lookDirection;
+
+        float_X parMomSinTheta;
+        float_X parMomCosTheta;
+        float_X const parMomPhi;
+        float_X parMomSinPhi;
+        float_X parMomCosPhi;
+        float_X detectorSinTheta;
+        float_X detectorCosTheta;
+        float_X const detectorPhi;
+        float_X const parSqrtOnePlusUSquared;
+
+    public: 
+        HDINLINE
+        Calculator(
+            transitionRadiation::Particle const & particleSet,
+            float3_X const lookDirection
+        ) : 
+            particle( particleSet ),
+            lookDirection( lookDirection ),
+            parMomPhi( particle.getMomPhi( ) ),
+            detectorPhi( 
+                picongpu::math::atan2(
+                    lookDirection.z( ), 
+                    lookDirection.x( )
+                ) + picongpu::PI
+            ),
+            parSqrtOnePlusUSquared( 
+                picongpu::math::sqrt( 1 + particle.getU( ) * particle.getU( ) ) 
+            )
+        { 
+            // Frequent calculations
+            // Momentum Space for Particle:
+            picongpu::math::sincos( 
+                particle.getMomTheta( ), 
+                parMomSinTheta, 
+                parMomCosTheta 
+            );
+            picongpu::math::sincos( 
+                parMomPhi - detectorPhi, 
+                parMomSinPhi, 
+                parMomCosPhi
+            );
+            
+            // Detector Position
+            float_X const detectorTheta = picongpu::math::acos( 
+                lookDirection.y( ) / picongpu::math::sqrt( ( 
+                    lookDirection * lookDirection
```

Comment:
Is your `lookDirection` not normalized?

Reply:
Since the lookdirection is set in the param files, it could theoretically be set by the user to a not normalized value... But I guess it should be clear from the comment, that `transtionradiation::observationDirection(const int observation_id_extern)` has to be normalized.